pyfar/dsp/regularization.py

```diff
@@ -0,0 +1,337 @@
+                "Regularization must be a pyfar.Signal object.")
+
+        if target is not None and not isinstance(target, pf.Signal):
+            raise ValueError(
+                "Target function must be a pyfar.Signal object.")
+
+        instance = cls(regu_type = "magnitude spectrum")
+        instance._regu = regularization
+        instance._target = target
+        instance._beta = beta
+        instance._normalize = normalize_regularization
+
+        return instance
+
+    def invert(self, signal):
+        """Invert the spectrum of a signal applying frequency dependent
+        regularization.
+
+        Parameters
+        ----------
+        signal : pf.Signal
+            Signal to be inverted.
+
+        Returns
+        -------
+        inverse : pf.Signal
+            Resulting signal after regularized inversion.
+        """
+
+        # get regularization factors
+        regu = self.regularization(signal)
+        regu_final = regu.freq
+
+        data = self._signal.freq
+
+        # calculate inverse filter
+        inverse = self._signal.copy()
+        if self._target is not None:
+            # Norcross 2006 eq. 2.13
+            inverse.freq = np.conj(data) * regu_final / (np.conj(data) * data)
+        else:
+            inverse.freq = np.conj(data) / (np.conj(data) * data + regu_final)
+
+        return inverse
+
+
+    def regularization(self, signal):
+        """
+        Get the frequency dependent regularization factors.
+
+        Parameters
+        ----------
+        signal : pf.Signal, optional
+            Signal on which the regularization will be used on. This is
+            nescessary to assure that the regularization factors have the same
+            number of bins as the signal.
+
+        Returns
+        -------
+        regu : pf.FrequencyData
+            Frequency dependent regularization as FrequencyData.
+        """
+
+        # zero pad, if target function is given
+        if self._target is not None and \
+            self._target.n_samples != signal.n_samples:
+            n_samples = max(signal.n_samples, self._target.n_samples)
+            signal = pf.dsp.pad_zeros(signal, (n_samples-signal.n_samples))
+            self._target = pf.dsp.pad_zeros(self._target,
+                                            (n_samples-self._target.n_samples))
+
+        # Assign signal to class attribute, so further processing methods can
+        # use zero padded signal (e.g. invert())
+        self._signal = signal
+
+        # Call private method to get regularization factors
+        if self._regu_type == "frequency range":
+            regu = self._get_regularization_from_frequency_range(signal)
+        elif self._regu_type == "magnitude spectrum":
+            regu = self._get_regularization_from_regularization_signal(signal)
+
+        # Normalize regularization factors to energy of signal
+        if self._normalize == 'energy':
+            regu.freq *= np.max(np.abs(signal.freq)**2)
```

Comment:
Two thoughts:

- The way this was used before, max(regu.freq)=1 was assumed. We now need to scale regu.freq by f(signal) / f(regu), where f is some funtion. 
- I would suggest to call the normalization 'maximum_energy' and introduce a second normalization termed 'energy' which uses f = np.abs()**2 without taking the maximum.